synapse/handlers/e2e_keys.py

```diff
@@ -981,8 +989,37 @@ def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None
             user_id, key_type, from_user_id
         )
         if key is None:
-            logger.debug("no %s key found for %s", key_type, user_id)
+            # Attempt to fetch the missing key from the remote user's server
+            user = UserID.from_string(user_id)
+            try:
+                remote_result = yield self.federation.query_client_keys(
+                    user.domain, {"device_keys": {user_id: []}}, timeout=10 * 1000
```

Comment:
Should the user ID we're querying be in here?

Reply:
I assume so? That's what the `user_id` in `{"device_keys": {user_id: []}}` is. I left the value as `[]` as according to [the MSC](https://github.com/uhoreg/matrix-doc/blob/cross-signing2/proposals/1756-cross-signing.md) that should query all of their keys?

> Does this save the key in the cache after fetching it and if not, should it?

It's not, currently, but I suspect we want to. However atm I don't see any cache involved here. It looks like it's just pulling them directly from the db.

Reply:
oh right yes, I was completely failing to read the object literal correctly.

and yeah, I think cache === database according to this code (https://github.com/matrix-org/synapse/blob/d2a9b45df018cc4a13558ebb127d7932430008c5/synapse/handlers/e2e_keys.py#L281). Either way if we are just intentionally not updating our local copy here then that's fine.

Reply:
Well, if there's other code that relies on these keys being available then we'd also need the same fix there, and in that case it'd be more efficient to update the database in each instance.

But, changing the state of the db here may break other assumptions, so there's an argument against it. Unfortunately I don't have a grasp on everything Synapse is doing related to cross-signing.

Reply:
Although, looking at what code calls `get_e2e_cross_signing_key`, it's only the sig upload servlet and federation key queries, both of which shouldn't be harmed by having the extra keys in there.

Let me see what the code for saving them after retrieving them looks like.

Reply:
I also wonder whether we need to notify clients of device updates.

Reply:
Looks plausible, but yes, unsure if you would need to notify clients.

Reply:
Probably best to notify clients.

Reply:
we seem to have a very similar situation in `E2eKeysHandler.query_devices`, which I don't think does any notification in this situation. After all, we're assuming that the user does not share any rooms with the remote user, so why would we notify clients of the new key.

Reply:
Indeed, I wonder if we should not be sharing code with `E2eKeysHandler.query_devices`

Reply:
I originally pulled `do_remote_query` out of that method to try and share it between `query_devices` and `_get_e2e_cross_signing_verify_key`, but decided it was doing too much and just plucked `self.federation.query_client_keys` out of it instead.

Now we've transitioned to `self.federation.query_user_devices` which is a lot simpler. I'm just now worried/wondering whether we should be taking the `device_keys` returned by this method and doing anything with them (which `query_devices` seems to do, but it does so much else as well).